File: utilities/shell/direnv-config/sdk/typescript/test/merge.test.ts
import { describe, it, expect } from 'vitest';
import { deepMerge, deepMergeMulti } from '../src/merge.js';

describe('deepMerge', () => {
  it('overlay scalar replaces base scalar', () => {
    expect(deepMerge('old', 'new')).toBe('new');
    expect(deepMerge(1, 2)).toBe(2);
  });

  it('overlay adds new keys to base map', () => {
    const result = deepMerge({ a: 1 }, { b: 2 });
    expect(result).toEqual({ a: 1, b: 2 });
  });

  it('maps merge recursively', () => {
    const base = { db: { host: 'localhost', port: 5432 } };
    const overlay = { db: { port: 3306, name: 'mydb' } };
    const result = deepMerge(base, overlay);
    expect(result).toEqual({ db: { host: 'localhost', port: 3306, name: 'mydb' } });
  });

  it('array from overlay replaces base array entirely', () => {
    const result = deepMerge({ items: [1, 2, 3] }, { items: [4, 5] });
    expect(result).toEqual({ items: [4, 5] });
  });

  it('type mismatch: overlay wins', () => {
    expect(deepMerge({ a: 'string' }, { a: 42 })).toEqual({ a: 42 });
    expect(deepMerge({ a: [1] }, { a: 'scalar' })).toEqual({ a: 'scalar' });
    expect(deepMerge('scalar', { a: 1 })).toEqual({ a: 1 });
  });

  it('tombstone (_dc_pruned: true) strips subtree', () => {
    const base = { keep: 1, remove: { nested: 'data' } };
    const overlay = { remove: { _dc_pruned: true } };
    const result = deepMerge(base, overlay);
    expect(result).toEqual({ keep: 1 });
  });

  it('nested tombstone strips only that branch', () => {
    const base = { a: { b: { c: 1 }, d: 2 } };
    const overlay = { a: { b: { _dc_pruned: true } } };
    const result = deepMerge(base, overlay);
    expect(result).toEqual({ a: { d: 2 } });
  });
});

describe('deepMergeMulti', () => {
  it('with empty array returns null', () => {
    expect(deepMergeMulti([])).toBeNull();
  });

  it('with single element returns it (stripped)', () => {
    const result = deepMergeMulti([{ a: 1, b: { _dc_pruned: true } }]);
    expect(result).toEqual({ a: 1 });
  });

  it('folds left-to-right correctly', () => {
    const layers = [
      { a: 1, b: 'base' },
      { b: 'middle', c: 3 },
      { c: 'top', d: 4 },
    ];
    const result = deepMergeMulti(layers);
    expect(result).toEqual({ a: 1, b: 'middle', c: 'top', d: 4 });
  });
});

File: sdk/elixir/test/direnv_config/merge_test.exs
defmodule DirenvConfig.MergeTest do
  use ExUnit.Case, async: true

  alias DirenvConfig.Merge

  describe "deep_merge/2" do
    test "overlay replaces scalar" do
      assert Merge.deep_merge(%{"a" => 1}, %{"a" => 2}) == %{"a" => 2}
    end

    test "adds new keys" do
      assert Merge.deep_merge(%{"a" => 1}, %{"b" => 2}) == %{"a" => 1, "b" => 2}
    end

    test "recursive map merge" do
      base = %{"db" => %{"host" => "localhost", "port" => 5432}}
      overlay = %{"db" => %{"port" => 3306, "name" => "mydb"}}

      assert Merge.deep_merge(base, overlay) == %{
               "db" => %{"host" => "localhost", "port" => 3306, "name" => "mydb"}
             }
    end

    test "list overlay replaces base" do
      base = %{"tags" => ["a", "b"]}
      overlay = %{"tags" => ["x"]}
      assert Merge.deep_merge(base, overlay) == %{"tags" => ["x"]}
    end

    test "non-map base is replaced by overlay" do
      assert Merge.deep_merge("old", %{"new" => true}) == %{"new" => true}
    end
  end

  describe "strip_tombstones/1" do
    test "tombstone strips subtree" do
      data = %{"a" => 1, "remove_me" => %{"_dc_pruned" => true}}
      assert Merge.strip_tombstones(data) == %{"a" => 1}
    end

    test "nested tombstone" do
      data = %{
        "top" => %{
          "keep" => "yes",
          "drop" => %{"_dc_pruned" => true}
        }
      }

      assert Merge.strip_tombstones(data) == %{"top" => %{"keep" => "yes"}}
    end

    test "top-level tombstone returns nil" do
      assert Merge.strip_tombstones(%{"_dc_pruned" => true}) == nil
    end

    test "tombstones in lists" do
      data = %{"items" => [%{"_dc_pruned" => true}, "kept"]}
      assert Merge.strip_tombstones(data) == %{"items" => [nil, "kept"]}
    end
  end

  describe "deep_merge_multi/1" do
    test "empty list returns nil" do
      assert Merge.deep_merge_multi([]) == nil
    end

    test "single element returns stripped copy" do
      input = %{"a" => 1, "b" => %{"_dc_pruned" => true}}
      assert Merge.deep_merge_multi([input]) == %{"a" => 1}
    end

    test "folds left-to-right" do
      layers = [
        %{"a" => 1, "b" => 2},
        %{"b" => 20, "c" => 3},
        %{"c" => 30, "d" => 4}
      ]

      assert Merge.deep_merge_multi(layers) == %{"a" => 1, "b" => 20, "c" => 30, "d" => 4}
    end

    test "multi merge with tombstones" do
      layers = [
        %{"a" => 1, "b" => %{"x" => 10}},
        %{"b" => %{"_dc_pruned" => true}, "c" => 3}
      ]

      assert Merge.deep_merge_multi(layers) == %{"a" => 1, "c" => 3}
    end
  end
end
